### PR TITLE
Add option to provide resource based on metric descriptor.

### LINKED ISSUE #230: Support multiple monitored resource types in one process
**NB:** Before opening a feature request against this repo, consider whether the feature should/could be implemented in exporter libraries in other languages. If so, please [open an issue on opencensus-specs](https://github.com/census-instrumentation/opencensus-specs/issues/new) first.

**Is your feature request related to a problem? Please describe.**
We are using `GetMonitoredResource` in our code to determine which monitoring resource type to use. The resource type is determined by the metric name currently. If the metric is a built-in metric, we send it with a non-`global` resource type. If not, we send it with `global` resource type as custom metrics.

Now `GetMonitoredResource` was removed in v0.12.7.

**Describe the solution you'd like**
Provide some option similar to `GetMonitoredResource`.

### PR DESCRIPTION
**Pull request by rghetia**
fixes #230

### CODE REVIEW

File: metrics.go [hunk 1]
```diff
@@ -126,8 +126,18 @@ func (se *statsExporter) metricToMpbTs(ctx context.Context, metric *metricdata.M
 	if metric == nil {
 		return nil, errNilMetricOrMetricDescriptor
 	}
-
-	resource := se.metricRscToMpbRsc(metric.Resource)
+	var resource *monitoredrespb.MonitoredResource
+	if get := se.o.ResourceByDescriptor; get != nil {
+		mr := get(&metric.Descriptor)
```
Comment (yanweiguo) [inline metrics.go]: What `metric.Descriptor` will be if I register the following:

```
requestCountM = stats.Int64(
       "request_count",
	"The number of requests",
	stats.UnitDimensionless)
view.Register(
		&view.View{
			Description: "The number of requests",
			Measure:     requestCountM,
			Aggregation: view.Count(),
			TagKeys: append(metrics.CommonRevisionKeys, metrics.PodTagKey, metrics.ContainerTagKey,
				metrics.ResponseCodeKey, metrics.ResponseCodeClassKey, metrics.NumTriesKey),
		})
```

Comment (rghetia) [reply metrics.go]:
```go
metric.Descriptor{
	Name:        "request_count",
	Description: "The number of requests",
	Unit:        metricdata.	UnitDimensionless,
	Type:        metric.TypeCumulativeInt64,
}
```

Comment (yanweiguo) [reply metrics.go]: Could we determine resource type by name? Name is more like an ID instead of the description. WDYT?

Comment (rghetia) [reply metrics.go]: actually Descriptor.Name = View.Name.

knative does determine resource type from name as per [this](https://github.com/knative/pkg/blob/f684a1078ac309f4b32e6240d9b622bea092296e/metrics/stackdriver_exporter.go#L130).
It seems you also need tags, can you confirm?

Comment (yanweiguo) [reply metrics.go]: Ah got it.

The tags are used to set the values of monitored resource labels. For example, `project_id`.

File: stackdriver.go [hunk 2]
```diff
@@ -255,6 +255,23 @@ type Options struct {
...
+	// based on the metric Descriptor. Most users will not need to set this,
+	// but should instead set ResourceDetector.
+	//
+	// The MonitoredResource and ResourceDetector fields are ignored if this
+	// field is set to a non-nil value.
+	//
+	// The ResourceByDescriptor is called to derive monitored resources from
+	// metric.Descriptor and label map associated with the metric. If any label is
+	// used for the derived resource then it will be removed the label map.
+	// Returned label map would then be a subset of the original map.
```
Comment (yanweiguo) [inline stackdriver.go]: Could you describe what the return label maps are used for?

File: metrics_test.go [hunk 3]
```diff
@@ -559,3 +560,313 @@ func TestMetricsToMonitoringMetrics_fromProtoPoint(t *testing.T) {
 		}
 	}
 }
+
+func TestGetMonitorResource(t *testing.T) {
```
Comment (yanweiguo) [inline metrics_test.go]: nit: Should the name be `TestResourceByDescriptor`?

Comment (rghetia) [reply metrics_test.go]: changed.

File: stackdriver.go (hunk 2 above)
Comment (rghetia) [reply stackdriver.go]: added the description.

**Comment by rghetia** [comment]
@yanweiguo PTAL.

**Comment by yanweiguo** [comment]
LGTM

**Comment by rghetia** [comment]
@bogdandrutu, @songy23 , @dinooliva can one of you please review this?

File: metrics.go [hunk 4]
```diff
@@ -154,12 +155,26 @@ func (se *statsExporter) metricToMpbTs(ctx context.Context, metric *metricdata.M
 			// TODO: (@rghetia) perhaps log this error from labels extraction, if non-nil.
 			continue
 		}
+
+		var rsc *monitoredrespb.MonitoredResource
+		var mr monitoredresource.Interface
+		if get := se.o.ResourceByDescriptor; get != nil {
```
Comment (songy23) [inline metrics.go]: Nit: why not just `se.o.ResourceByDescriptor != nil`?

Comment (rghetia) [reply metrics.go]: done.

**Pull request merged**